### DISCUSSION TITLE
Attribute node(s)

### DISCUSSION THREAD

**Discussion by Marcus-Richmond** [Marcus-Richmond/sverchok-gis-nodes#14] [Ideas]
I've been thinking more about how to import attribute(s) of gis layers and I think I have a good solution - create two nodes:

1. Import Attributes as Dictionary node
    - This node would pull in all attributes from a layer and output to Sverchok as a dictionary.
2. Import Attribute as List node
    - This node would extract a single attribute from a layer and output to Sverchok as a list.

I was confused as to why the Import Attribute node was outputting a dictionary. Shown [here ](https://github.com/Marcus-Richmond/sverchok-gis-nodes/pull/8#issuecomment-1100669415)and below: 

![image](https://user-images.githubusercontent.com/59568811/164143692-5448fad1-bd33-4d79-bedc-216c13e366e9.png)

And I saw that the Import Attribute node has this section of code:

```
if DEBUG and path:

        if layername:
            gpd1 = gpd.read_file(path, layer=layername)
            gi = gpd1.__geo_interface__
        else:
            gpd1 = gpd.read_file(path)
            gi = gpd1.__geo_interface__
        
        if not hasattr(sverchok, 'gis_breakpoint'):
            sverchok.gis_breakpoint = {}
        
        sverchok.gis_breakpoint = gi
        listAttribute = [gi]
```

Which assigns `gi` (dictionary) to the output. But, the original way I thought this attribute node would be most useful is in this part:

```
if path and layername and attribute:

      # read in gpkg layer as geopandas data frame, and make alias
      gpd1 = gpd.read_file(path, layer=layername)
      gi = gpd1.__geo_interface__

      # loop over gi to extract and collect attributes from feature properties.
      for features in range(len(gi['features'])):
          value = gi['features'][features]['properties'][attribute]
          listAttribute.append(value)
```

Which adds the attribute from each feature to a list, and outputs the list to Sverchok. I also originally was unsure if an attribute loader should handle multiple attributes at once or just deal with a single attribute. I started with a single attribute since I found it easier to code, but now I'm really seeing the utility of having the geo interface dictionary output directly to Sverchok. This definitely seems like the best way to handle multiple attributes at once.

Anyway, thinking aloud here. I'm going to work on these Attribute Loader nodes.

**Comment by Marcus-Richmond** [comment]
Also, test data can be found here: [https://github.com/Marcus-Richmond/example-datasets/blob/main/TestShapes.zip](https://github.com/Marcus-Richmond/example-datasets/blob/main/TestShapes.zip)

**Comment by zeffii** [comment]
our problem will not be "how to code it" but "what to code."

during development numerous nodes will be introduced which later on may turn out to be too specific, and don't generalize well. This is fine, the goal is to refine and keep refining until you can't shave off anything anymore.

it may be that the most resource-efficient way to handle large datasets is to code a scriptnode by hand (which is probably not what die-hard node heads like to hear..) or including a scattering of smaller scripts throughout the tree.

- I think we can still make cool stuff, but we won't get it all perfect from the start.

**Comment by zeffii** [comment]
![image](https://user-images.githubusercontent.com/619340/164319764-12bbca55-ac81-4aab-a4c1-dc68d19fbbf9.png)

**Comment by Marcus-Richmond** [comment]
I'm having some trouble understanding how dictionaries work in Sverchok, for example I tried writing this code to use in the scripted node:

```
"""
in gpkg_path FP
in gpkg_layer s
out attribute_dict s
"""

import geopandas as gpd
import pandas as pd
import fiona
import numpy as np

attribute_dict = {}

if gpkg_path and gpkg_layer:
    
    # sanity check
    print(gpkg_path)
    
    # read in gpkg layer as geopandas data frame
    gpd1 = gpd.read_file(gpkg_path[0][0], layer = gpkg_layer[0][0])
    
    df = pd.DataFrame(gpd1.drop(columns='geometry'))
    
    attribute_dict = df.to_dict(orient='list')
    
else:
    print("not executed yet because no filepath and/or layer was given")
```
But, with various setups I can't get this working:

![image](https://user-images.githubusercontent.com/59568811/164369704-7c719d80-1f68-448e-abd1-99c05f3669c0.png)

![image](https://user-images.githubusercontent.com/59568811/164369790-aad89d0c-ef96-417d-8b6b-fffbb3b07578.png)

Any ideas? I was really glad to find the .to_dict method in pandas, which produces exactly the kind of dictionary I think will be most useful.

**Comment by zeffii** [comment]
- The error about recursion levels comes from the dictionary out node - I've never seen this error before - i'll look into it later.
- When i suggested the dictionary socket I hadn't ever used it before. Ultimately it may be what we want, but perhaps using a SvStringsSocket (the green socket) is less complicated to start off with, because we can deal with the data exactly as it comes in (so no extra interface to the dictionary provided by SvDict class)

**Reply by Marcus-Richmond** [reply]
Thanks, I'm also trying to figure out this recursion error and how to get the dictionary working. I found something even weirder - the recursion error seems to be triggered when strings exist as the values in a dict:

![image](https://user-images.githubusercontent.com/59568811/164584733-ac0663a7-27b0-4b38-82cc-a389e85ae5a4.png)

Maybe has to do with how geopandas writes this dictionary?

Some references:
- [https://datascience.stackexchange.com/questions/28868/attributeerror-str-object-has-no-attribute-keys](https://datascience.stackexchange.com/questions/28868/attributeerror-str-object-has-no-attribute-keys)
- [https://stackoverflow.com/questions/56121561/pythons-json-attributeerror-str-object-has-no-attribute-keys](https://stackoverflow.com/questions/56121561/pythons-json-attributeerror-str-object-has-no-attribute-keys)
- [https://stackoverflow.com/questions/48117272/python-keyerror-when-calling-valid-key-index-in-dict](https://stackoverflow.com/questions/48117272/python-keyerror-when-calling-valid-key-index-in-dict)

**Reply by zeffii** [reply]
this is interesting, nice bit of detective work :)   i will also meditate on this.

**Reply by zeffii** [reply]
it's even weirder.  If you create a copy of part of the node tree that uses a functioning dict_out node,  the Node doesn't seem to care.

![image](https://user-images.githubusercontent.com/619340/164652501-e6919f50-bd86-46d9-a032-12aca8f50154.png)

**Reply by zeffii** [reply]
right now the SvDict class is not doing what I hoped it would be able to do automatically, it is however probably behaving exactly as intended. 
- we know it can automatically make a dict (SvDict)
- we know now there are (artificial) limitations.

**Reply by Marcus-Richmond** [reply]
@zeffii Any chance you found a resolution to this error? Trying to get back into this, starting with attributes and importing them as a dict, which I still think is a good way to do it, but also still can't get past this error.

**Reply by Marcus-Richmond** [reply]
Here is something interesting - I tried comparing your script [below](https://github.com/Marcus-Richmond/sverchok-gis-nodes/discussions/14#discussioncomment-2614661):

```
"""
in gis_path FP
out main_dict s
"""

try:
    import geopandas as gpd

    dataset = gpd.datasets.get_path('naturalearth_lowres')
    world_gdf = gpd.read_file(dataset)
    geom = world_gdf.to_dict()
    main_dict.append(geom)
    
    
except Exception as err:
    print(err)
```

with the attribute import script I included [above](https://github.com/Marcus-Richmond/sverchok-gis-nodes/discussions/14#discussioncomment-2606040), which was giving the recursion error

```

"""
in gpkg_path FP
in gpkg_layer s
out attribute_dict s
"""

import geopandas as gpd
import pandas as pd
import fiona
import numpy as np

attribute_dict = {}

if gpkg_path and gpkg_layer:
    
    # sanity check
    print(gpkg_path)
    
    # read in gpkg layer as geopandas data frame
    gpd1 = gpd.read_file(gpkg_path[0][0], layer = gpkg_layer[0][0])
    
    df = pd.DataFrame(gpd1.drop(columns='geometry'))
    
    attribute_dict = df.to_dict(orient='list')
    
else:
    print("not executed yet because no filepath and/or layer was given")
```
noticing that the `.to_dict` method has different parameters between these two scripts, so I modified my script to remove the `orient='list'` parameter and:

![image](https://user-images.githubusercontent.com/59568811/178113857-0494d096-1997-4429-8573-e4af1e88fa34.png)

it successfully removed the error!

so, seems that the `orient='list'` causes the error?

**Reply by Marcus-Richmond** [reply]
Also, thanks for making the get keys and values node, I almost forgot you added this. This is so sweet:

![image](https://user-images.githubusercontent.com/59568811/178114267-285963dd-9a2b-453e-9128-10f963be8a1f.png)

**Reply by Marcus-Richmond** [reply]
I'm thinking next step I'd like to tackle is getting the attribute dictionary into a node, as well as create a node that just outputs a single attribute.

**Reply by zeffii** [reply]
yes! by all means do it, and if you run into issues ping me.

**Reply by Marcus-Richmond** [reply]
Yep, just added the attribute as dict and single attribute as list nodes, will probably organize them into a separate nodes menu as well.

**Comment by zeffii** [comment]
about the script node script:

```
"""
in gpkg_path FP
in gpkg_layer s
out attribute_dict s           < ------ this will be available as a List in the code below
"""

import geopandas as gpd
import pandas as pd
import fiona
import numpy as np

attribute_dict = {}              <----- here you are overwriting a local variable from a list to a dict

if gpkg_path and gpkg_layer:
    
    # sanity check
    print(gpkg_path)
    
    # read in gpkg layer as geopandas data frame
    gpd1 = gpd.read_file(gpkg_path[0][0], layer = gpkg_layer[0][0])
    
    df = pd.DataFrame(gpd1.drop(columns='geometry'))
    
    attribute_dict = df.to_dict(orient='list')           # <------------ overwriting it again.
    
else:
    print("not executed yet because no filepath and/or layer was given")
```
instead all you would need to write is.
```python
"""
in gpkg_path FP
in gpkg_layer s
out attribute_dict s
"""

import geopandas as gpd
import pandas as pd
import fiona
import numpy as np

if gpkg_path and gpkg_layer:
    # print(gpkg_path)
    
    # read in gpkg layer as geopandas data frame
    gpd1 = gpd.read_file(gpkg_path[0][0], layer = gpkg_layer[0][0])
    df = pd.DataFrame(gpd1.drop(columns='geometry'))
    
    attribute_dict.append(df.to_dict(orient='list'))
    
else:
    print("not executed yet because no filepath and/or layer was given")
```
the output of the socket would then be `[{}]`  , a single dict inside a List.  Think back to the double nesting, the outer List is merely the container to allow the option to include multiple dicts to be passed around in sockets

**Comment by zeffii** [comment]
@Marcus-Richmond  so we might be 
- passing around structures like `gpd1` and 
- have nodes specifically to peform `pd.DataFrame(gpd1.drop.....)` 
    - with the option to output as dict using df.to_dict with whatever options that function provides

in essence turning certain functions of these libraries into Nodes which can be strung together, and that allows you to consider nodes as if they were a function.  Easy to maintain functions are short functions, avoid trying to do too much inside a single node.

**Reply by zeffii** [reply]
altho, in Sverchok we have a great many nodes that ignore the principle of simplicity (for good reason).

**Comment by zeffii** [comment]
with the backend settled, i'll take time to read more into gpd/pd/fiona - there's some really interesting stuff there.

**Comment by zeffii** [comment]
![image](https://user-images.githubusercontent.com/619340/164684237-225080df-b11b-4b17-b320-cbad41baf9cf.png)

```python
"""
in gis_path FP
out main_dict s
"""

try:
    import geopandas as gpd

    dataset = gpd.datasets.get_path('naturalearth_lowres')
    world_gdf = gpd.read_file(dataset)
    geom = world_gdf.to_dict()
    main_dict.append(geom)
    
    
except Exception as err:
    print(err)
```

**Comment by zeffii** [comment]
i've added the dictionary socket type to SNLite. it uses `D`.   (you'll need to update Sverchok )

```python
"""
in gis_path FP
out main_dict D
"""

try:
    import geopandas as gpd

    dataset = gpd.datasets.get_path('naturalearth_lowres')
    world_gdf = gpd.read_file(dataset)
    geom = world_gdf.to_dict()
    main_dict.append(geom)
    
    
except Exception as err:
    print(err)
```
![image](https://user-images.githubusercontent.com/619340/164688061-8d2a51fb-17cb-4404-a958-89e2521b2221.png)

**Reply by zeffii** [reply]
![image](https://user-images.githubusercontent.com/619340/164700234-a57583ee-42c7-4d84-b6ac-fd680306ad42.png)

**Reply by zeffii** [reply]
![image](https://user-images.githubusercontent.com/619340/164702857-92a2f073-f900-4def-b60e-c76a0e70331d.png)

**Reply by zeffii** [reply]
will need to revisit multipolygon extraction :) 
```python
"""
in geom s
out verts v
out edges s
"""

if geom and geom[0]:
    for geo in geom[0]:
        coords = geo.__geo_interface__['coordinates']
        geo_type_str = str(geo.type)
        if geo_type_str == "Polygon":   # don't do this.
            array_2d = np.array(coords[0])
            array_3d = np.c_[ array_2d, np.zeros(array_2d.shape[0])]
            verts.append(array_3d)
        if geo_type_str == "MultiPolygon":   # don't do this.

            temp_array = None # np.array()
            for polyline in coords:
                array_2d = np.array(polyline[0])
                array_3d = np.c_[ array_2d, np.zeros(array_2d.shape[0])]
                if not isinstance(temp_array, np.ndarray):
                    temp_array = array_3d
                else:
                    np.append(temp_array, array_3d, 0)
    
            verts.append(temp_array)
```
![image](https://user-images.githubusercontent.com/619340/164719288-65645eaa-5694-4bfd-b617-173e4ed0055b.png)

**Reply by zeffii** [reply]
ahh.. a multipolygon can be at least partially dissolved using
```
merged = list(pandas.core.common.flatten(list2d))

```
numpy can then reshape it :)  `reshape((-1, 2))`

**Reply by zeffii** [reply]
```python
"""
in geom s
out verts v
out edges s
"""
try:
    import pandas as pd

    if geom and geom[0]:
        for geo in geom[0]:
            coords = geo.__geo_interface__['coordinates']
            geo_type_str = str(geo.type)
            if geo_type_str == "Polygon":   # don't do this.
                array_2d = np.array(coords[0])
                array_3d = np.c_[ array_2d, np.zeros(array_2d.shape[0])]
                verts.append(array_3d)
            if geo_type_str == "MultiPolygon":   # don't do this.
                array_2d = np.array(list(pd.core.common.flatten(coords))).reshape((-1, 2))
                array_3d = np.c_[ array_2d, np.zeros(array_2d.shape[0])]
                verts.append(array_3d)

except Exception as err:
    print(err)
```

**Reply by zeffii** [reply]
the above script is written more sparsely like:
```python
"""
>in geom s
out verts v
out edges s
"""
try:
    import pandas as pd

    for geo in geom[0]:
        coords = geo.__geo_interface__['coordinates']
        geo_type_str = str(geo.type)

        if geo_type_str == "Polygon":   # don't do this.
            array_2d = np.array(coords[0])
        elif geo_type_str == "MultiPolygon":   # don't do this.
            array_2d = np.array(list(pd.core.common.flatten(coords))).reshape((-1, 2))
        else:
            continue

        verts.append(np.c_[ array_2d, np.zeros(array_2d.shape[0])])

except Exception as err:
    print(err)
```

**Reply by zeffii** [reply]
seen as shapely is a useful dependency, there's sense to use `isinstance(geo, <the exact type>)`
```python
"""
>in geom s
out verts v
out edges s
"""
try:
    import pandas as pd
    import shapely.geometry as sg

    for geo in geom[0]:
        coords = geo.__geo_interface__['coordinates']

        if isinstance(geo, sg.Polygon):
            array_2d = np.array(coords[0])
        elif isinstance(geo, sg.MultiPolygon):
            array_2d = np.array(list(pd.core.common.flatten(coords))).reshape((-1, 2))
        else:
            continue

        verts.append(np.c_[ array_2d, np.zeros(array_2d.shape[0])])

except Exception as err:
    print(err)
```

building edges will be a little less elegant, unless shapely offers a convenient method to extract the data

**Reply by zeffii** [reply]
but https://shapely.readthedocs.io/en/stable/manual.html#numpy-and-python-arrays that this can be rewritten
```python
array_2d = np.array(coords[0])
vs
array_2d = np.asarray(geo)
```
but i doubt much efficiency is to be expected in the individual polygon/multipolygon scenario .

**Comment by zeffii** [comment]
```python
"""
>in geom s
out verts v
out edges s
"""
try:
    import pandas as pd
    import shapely.geometry as sg
    from sverchok.data_structure import get_edge_loop

    for geo in geom[0]:
        coords = geo.__geo_interface__['coordinates']

        if isinstance(geo, sg.Polygon):
            array_2d = np.array(coords[0])
            edge_list = get_edge_loop(array_2d.shape[0])
        elif isinstance(geo, sg.MultiPolygon):
            # array_2d = np.array(list(pd.core.common.flatten(coords))).reshape((-1, 2))
            continue
        else:
            continue
        edges.append(edge_list)
        verts.append(np.c_[ array_2d, np.zeros(array_2d.shape[0])])

except Exception as err:
    print(err)
```
### starting on edges
for single series edgeloops, `sverchok.data_structure` has a function
just a matter of finding time to code something similar for MultiPolygon to edges.
Maybe it's not obvious from the code, but the multipolygon is an `iterable of iterables of iterables of ....`  , and because all i want are the Points, it's possible to flatten the nested iterables into a single dimensional list, then reshape it to pairs of coordinates.
